File: Episode1/LogisticRegression.jl
using CSV, DataFrames, LinearAlgebra, Plots, BenchmarkTools

mutable struct LinearModel
    step_size::Float64
    max_iter::Int64
    eps::Float64
    theta::Vector{Float64}
    verbose::Bool
end

function LinearModel(theta::Vector{Float64}=zeros(1),step_size::Float64=0.2, max_iter::Int64=100, eps::Float64=1e-5, verbose::Bool=true)

    return LinearModel(step_size,max_iter,eps,theta,verbose)

end

function sigm(x)
    return 1/(1+exp(-x))
end

function fit(model::LinearModel, x,y)

    m,n=size(x)
    model.theta = zeros(n)
    
    i = 1
    while true
        oldtheta = copy(model.theta)
        ht = sigm.(-x*model.theta)
        H = H = 1/m * x' * diagm(ht .* (1 .- ht)) * x
        gradJ = 1/m *x'*(y.-ht)
        model.theta -= H \ gradJ

        i += 1
        if norm(model.theta - oldtheta, 1) < model.eps
            break
        end 
    end
end

function plt(x,y)
    scatter!(x[:, 2], x[:, 3], zcolor=y)
end

function pltfit(model::LinearModel, x, y)
    x1 = minimum(x[:,2])
    x2 = maximum(x[:,2])
    plot!([x1,x2], [-(x1*model.theta[2]+model.theta[1])/model.theta[3],-(x2*model.theta[2]+model.theta[1])/model.theta[3]])
end

scatter()
model = LinearModel()

function csv_to_matrix(filename::String)
    # Read the CSV file into a DataFrame
    df = CSV.read(filename, DataFrame)
    n = size(df[1,:])[1]

    x = Matrix(hcat(ones(size(df, 1)), df[:, 1:end-1])) 
    y = Vector(df[:,n])
    return (x,y)
end

# Example usage:
filename = "file_name.csv" # Replace with your CSV filename
x,y = csv_to_matrix(filename)

plt(x,y)

# Only benchmarks the fit function
@btime fitLOGREG(model,x,y)

pltfit(model,x,y)
savefig("my_plot.png") 

File: Episode1/LogisticRegression.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from time import *


class LinearModel(object):
    def __init__(self, step_size=0.2, max_iter=100, eps=1e-5,
                 theta=None, verbose=True):
        
        """
        Args:
            step_size: Step size for iterative solvers only.
            max_iter: Maximum number of iterations for the solver.
            eps: Threshold for determining convergence.
            theta_0: Initial guess for theta. If None, use the zero vector.
            verbose: Print loss values during training.
        """
        self.theta = theta
        self.step_size = step_size
        self.max_iter = max_iter
        self.eps = eps
        self.verbose = verbose

    
    def fit(self, x, y):
        """
        Args:
            x: Training example inputs. Shape (m, n).
            y: Training example labels. Shape (m,).
        """
        raise NotImplementedError('Subclass of LinearModel must implement fit method.')


class LogisticRegression(LinearModel):

    def fit(self, x, y):
        """
        Args:
            x: Training example inputs. Shape (m, n).
            y: Training example labels. Shape (m,).
        """
        m,n=np.shape(x)
        self.theta = np.zeros((n,1)).flatten()

        while True:
            oldtheta = np.copy(self.theta)
            ht = 1/(1+np.exp(-x.dot(self.theta)))
            gradJ = 1/m*(x.T).dot(y - ht)
            H = (1/m) * x.T.dot(np.diagflat(ht*(1-ht))).dot(x)
            self.theta -= -np.linalg.solve(H, gradJ)

            if np.linalg.norm(self.theta-oldtheta,ord=1) < self.eps:
                break

    def plotfit(self, x, y):
        x1 = x[:,1]
        x2 = x[:,2]
        plt.scatter(x1,x2,c=y)
        inp = np.linspace(x1.min(),x1.max(),100)
        out = -(self.theta[1]*inp+self.theta[0])/self.theta[2]
        plt.plot(inp,out,"-r")
        plt.show()


def csv_to_matrix(filename):
  """
  Converts a CSV file into a NumPy matrix.

  Args:
    filename: Path to the CSV file.

  Returns:
    A NumPy matrix representing the data from the CSV file.
  """

  # Read the CSV file into a Pandas DataFrame
  df = pd.read_csv(filename)

  # Convert the DataFrame to a NumPy matrix
  matrix = df.to_numpy()

  return matrix

# Example usage:
filename = 'file_name.csv'
data_matrix = csv_to_matrix(filename)
n = np.shape(data_matrix)[1]
x = data_matrix[:,0:n-1]
y = data_matrix[:,n-1]

# Adding a column of 1s to the matrix x
x = np.insert(x, 0, 1, axis=1) 

###
clf = LogisticRegression()

# Manual benchmarking
store = 0
n = 100
for i in range(n):
    start = perf_counter()
    clf.fit(x, y)
    end = perf_counter()
    store += end - start

print(store/n)

clf.plotfit(x,y)